Implement GT_NEG decomposition

mikedn opened the pull request:
Fixes #4171

cmckinsey commented:
Love to see you taking this. Do you want to self-assign #4171 so Michelle doesn't accidentally start working on it?

mikedn commented:
@cmckinsey I don't think I can self-assign GitHub issues, you or someone else from MS has to assign it to me. Same for #4659 and #4593.

cmckinsey commented:
@mikedn  Oh right. I'm happy to do it. Whenever you want to grab an item, just post a note on the issue and someone on the team @RussKeldorph will assign to you so we avoid duplicate work.

RussKeldorph commented:
@cmckinsey I'm not sure which cred allows you to be assigned issues (.NET Foundation member, maybe?), but we can't assign to @mikedn.  In lieu of that, I think the best we can do is label the issue "grabbed by community" and have a comment acknowledging the owner.

mikedn commented:
This one should be good to go after #5698 is fixed, then I can remove the "temporary workarounds" commit.

mikedn commented:
@cmckinsey Thanks for the membership invitation! Looks like now I can be assigned issues.

mikedn commented:
Now that other x86 issues have been fixed this is ready for review.

sivarv commented:
CC @dotnet/jit-contrib

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -491,7 +491,63 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+
+            GenTree* loNeg = comp->gtNewOperNode(GT_NEG, TYP_INT, loOp1);
+            loNeg->CopyCosts(tree);
+            loNeg->gtFlags = (tree->gtFlags & GTF_LIVENESS_MASK);
```
I'm not sure why the `GTF_LIVENESS_MASK` is being copied here, as a `GT_NEG` should never have those bits set - though I see that `DecomposeStoreInd` is also doing that, in that case for a `GT_IND`.

mikedn replied:
Ah, I suppose those flags are used only on lclvar nodes? You suggested copying those flags in another [PR](https://github.com/dotnet/coreclr/pull/4986#discussion_r63386749) but now realize that it was specific to `GT_STORE_LCL_VAR` decomposition.

BruceForstall replied:
Yeah, `GTF_LIVENESS_MASK` only applies to lclVar nodes (it's too bad we don't have better debugging modes to detect reading/writing bits you're not supposed to touch, based on node type). It's probably ok to copy costs, although as you point out, it might not matter.

---
In reply to: [70025072](https://github.com/dotnet/coreclr/pull/5966#discussion_r70025072) [](ancestors = 70025072)

mikedn replied:
OK, I'll remove `GTF_LIVENESS_MASK` and double check if there's nothing else that needs to be copied. What I know for sure is that I must not copy the side effects. They also need to be removed from the original `GT_NEG` due to the introduction of temporaries.

CarolEidt commented:
LGTM apart from one confusing thing. @sejongoh, @adiaaida and/or @BruceForstall could you take a look and comment on the `GTF_LIVENESS_MASK` question?

mikedn commented on src/jit/lower.cpp:
```diff
@@ -491,7 +491,63 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+            //
+
+            GenTree* loNeg = comp->gtNewOperNode(GT_NEG, TYP_INT, loOp1);
+            loNeg->CopyCosts(tree);
```
And speaking of copying/not copying... do we actually need to copy costs? Decomposition doesn't do it consistently and the only post-lower use of costs I could find is some debug code in codegen. And event that code only looks at statement costs and ignores all other nodes.

BruceForstall commented on src/jit/lower.cpp:
```diff
@@ -491,7 +491,63 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+            }
+
+            newStmt = comp->fgInsertEmbeddedFormTemp(&op1->gtOp.gtOp2);
+            assert(!newStmt->gtStmtIsTopLevel());  
```
I'm wondering why this case of `fgInsertEmbeddedFormTemp` doesn't also need the logic from above to recurse through the possible embedded statements in case a new top-level stmt is created. Is there some reason the case handled there won't occur in this case? I agree with the "TODO-Cleanup" note!

mikedn replied:
That logic is only needed if the newly added statement is a top level statement instead of an embedded statement. This can only happen if the tree that's stored in a temporary starts right at the beginning of the current statement. So, once a temporary is created there's no way that subsequent temporary creation could produce a new top level statement, the `GT_LCL_VAR` node created for the first temporary is in the way.

There's an assert there to make sure my understanding of embedded statements is correct, I've been bitten to many times by embedded statement issues.

You can see similar code in other lowering code, `LowerArrElem` for example. Once we create a function the same logic will probably always be used, event though it's unnecessary in some cases.

BruceForstall commented:
LGTM after the GTF_LIVENESS_MASK fix. I didn't look to see if we need to propagate any other bits.

BruceForstall commented:
Sorry, you'll have a non-trivial merge after https://github.com/dotnet/coreclr/commit/c99eaa4d0e3ff619e98b410af5fb619cebaa58b4

mikedn commented:
No problem, thanks for refactoring!

mikedn commented:
Rebased and updated:
* Extended an assert in `LowerArg` to cover `GT_NEG` in addition to `GT_ADD` and `GT_SUB`.
* No more `CopyCosts`. Not only that costs are not used post lowering but they're wrong too. As far as I can tell costs are tree properties, not node properties. If you're doing things like temporary insertion then the costs no longer reflect the modified tree.
* Copy all flags (after clearing side effects). As far as I can tell there's no need to copy anything but it seems safer to just copy everything as long the copied flags are correct.
* Modified the way the hi tree is created and inserted so I can use `FinalizeDecomposition`.
* And of course, used `CreateTemporary`.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -905,6 +905,56 @@ void DecomposeLongs::DecomposeNot(GenTree** ppTree, Compiler::fgWalkData* data)
 
+//------------------------------------------------------------------------
+// DecomposeNot: Decompose GT_NEG.
+//
```
cut-and-paste error: => DecomposeNeg

BruceForstall commented:
LGTM. Thanks for all the great work on this, and decomposition in general.

Do you need to kick the CI to rerun the testing?

mikedn commented:
Fixed cut-and-past error.
